File: 15-Flyweight/src/go/tree_factory.go
package main

import "fmt"

// TreeFactory creates and caches shared *TreeType flyweights.
// The same intrinsic combination always returns the SAME pointer.
type TreeFactory struct {
	cache map[string]*TreeType
}

func NewTreeFactory() *TreeFactory {
	return &TreeFactory{cache: make(map[string]*TreeType)}
}

func (f *TreeFactory) GetTreeType(name, color, texture string) *TreeType {
	key := name + "-" + color + "-" + texture
	if t, ok := f.cache[key]; ok {
		return t
	}
	fmt.Printf("  [Factory] Creating new TreeType: %s\n", key)
	t := &TreeType{name: name, color: color, texture: texture}
	f.cache[key] = t
	return t
}

func (f *TreeFactory) DistinctTypeCount() int { return len(f.cache) }

File: 15-Flyweight/src/rust/tree_factory.rs
use std::collections::HashMap;
use std::rc::Rc;

use crate::tree_type::TreeType;

/// Flyweight Factory — creates and caches shared `Rc<TreeType>` flyweights.
/// The same intrinsic combination always returns a clone of the SAME `Rc`
/// (a refcount bump, not a data copy).
pub struct TreeFactory {
    cache: HashMap<String, Rc<TreeType>>,
}

impl TreeFactory {
    pub fn new() -> Self {
        TreeFactory { cache: HashMap::new() }
    }

    pub fn get_tree_type(&mut self, name: &str, color: &str, texture: &str) -> Rc<TreeType> {
        let key = format!("{}-{}-{}", name, color, texture);
        if let Some(t) = self.cache.get(&key) {
            return Rc::clone(t);
        }
        println!("  [Factory] Creating new TreeType: {}", key);
        let t = Rc::new(TreeType::new(name.to_string(), color.to_string(), texture.to_string()));
        self.cache.insert(key, Rc::clone(&t));
        t
    }

    pub fn distinct_type_count(&self) -> usize {
        self.cache.len()
    }
}
